### PR TITLE
various forward declarations

### PR DESCRIPTION
**Pull request by sjunges**
Various forward declarations to speed up compiling and to prevent recompilation. Also updates the carl-repo (but should not break support for old repo right now).

I have not yet updated the github actions because I want this to run through with the old version of carl.

### CODE REVIEW

**Comment by sjunges** [comment]
It is really really inconvenient that the necessary standard headers are system-dependent...

File: resources/3rdparty/CMakeLists.txt [hunk 1]
```diff
@@ -378,6 +381,18 @@ if (STORM_SHIPPED_CARL)
     endif()
 endif()
 
+if("${carl_VERSION_MAJOR}.${carl_VERSION_MINOR}" VERSION_EQUAL "14.22")
+    # This version is too old for forward declarations and updating requires moving the git,
+    # so we warn users but start warning them now.
+    set(STORM_CARL_SUPPORTS_FWD_DECL OFF)
+    message(WARNING "Uses an outdated repo for Carl. Carl is now hosted at ")
```
Comment (volkm) [inline resources/3rdparty/CMakeLists.txt:388]: We should add the new host location `https://github.com/moves-rwth/carl-storm`

File: src/storm/logic/BinaryPathFormula.cpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (volkm) [inline src/storm/logic/BinaryPathFormula.cpp:1]: Remove empty line

**Review by volkm — COMMENTED** [review]
A couple minor things, but looks good overall. Thanks for all the work.

File: Dockerfile [hunk 3]
```diff
@@ -19,7 +19,7 @@ ARG no_threads=1
 WORKDIR /opt/
 
 # Obtain Carl from public repository
-RUN git clone -b master14 https://github.com/ths-rwth/carl.git
+RUN git clone https://github.com/moves-rwth/carl.git
```
[the remark this answers is not in the document]
Comment (sjunges) [reply Dockerfile:22]: Yes, thanks!

**Comment by sjunges** [comment]
Thanks for your review @volkm. I also updated the CI. It seems to work with the new github repo

**Comment by volkm** [comment]
Looks good. Feel free to merge it.

**Comment by sjunges** [comment]
I am rerunning this with carl 14.24, if this passes I will merge.

**Pull request merged**